src/Strategy/Cargo.hs

```diff
@@ -383,8 +402,109 @@ buildGraph meta = stripRoot $
     traverse_ direct $ metadataWorkspaceMembers meta
     traverse_ addEdge $ resolvedNodes $ metadataResolve meta
 
-parsePkgId :: Text.Text -> Parser PackageId
-parsePkgId t =
+-- | Custom Parsec type alias
+type PkgSpecParser a = Parsec Void Text a
+
+-- | Parser for pre cargo v1.77.0 package ids.
+oldPkgIdParser :: Text -> Either Text PackageId
+oldPkgIdParser t =
   case Text.splitOn " " t of
-    [a, b, c] -> pure $ PackageId a b c
-    _ -> fail "malformed Package ID"
+    [a, b, c] -> Right $ PackageId a b c
+    _ -> Left $ "malformed Package ID: " <> t
+
+type PkgName = Text
+type PkgVersion = Text
+
+parsePkgSpec :: PkgSpecParser PackageId
+parsePkgSpec = eatSpaces (try longSpec <|> simplePkgSpec')
+  where
+    eatSpaces m = space *> m <* space
+
+    -- In the following imagine we're working with this fragment:
+    -- adler@1.0.2
+    pkgName :: PkgSpecParser (PkgName, PkgVersion)
+    pkgName = do
+      -- Parse: adler
+      name <- takeWhile1P (Just "Package name") (`notElem` ['@', ':'])
+      -- Parse: @1.0.2
+      version <- optional (choice [char '@', char ':'] *> semver)
+      -- It's possible to specify a name with no version, use "*" in this case.
+      pure (name, fromMaybe "*" version)
+
+    simplePkgSpec' =
+      pkgName >>= \(name, version) ->
+        pure
+          PackageId
+            { pkgIdName = name
+            , pkgIdVersion = version
+            , pkgIdSource = ""
+            }
+
+    -- In the following imagine we are working with this spec:
+    -- registry+https://github.com/rust-lang/crates.io-index#adler@1.0.2
+    longSpec :: PkgSpecParser PackageId
+    longSpec = do
+      -- Parse: registry+https
+      sourceInit <- takeWhile1P (Just "Initial URL") (/= ':')
+      -- Parse: ://github.com/rust-lang/crates.io-index
+      sourceRemaining <- takeWhile1P (Just "Remaining URL") (/= '#')
+      let pkgSource = sourceInit <> sourceRemaining
+
+      -- In cases where we can't find a real name, use text after the last slash as a name.
+      -- Or as a very last resort the entire package source path.
```

Comment:
[Optional]: Provide explicit result from the example 

```suggestion
      -- In cases where we can't find a real name, use text after the last slash as a name. (i.e. ...)
      -- Or as a very last resort the entire package source path. (i.e. ...)
```

Reply:
Good idea, I've added that.